mono/metadata/superluminal.c

```diff
@@ -0,0 +1,330 @@
+	enumerationData->mNumAssemblies++;
+
+	MonoImage *image = mono_assembly_get_image (assembly);
+	image_event (image, TRUE, FALSE);
+}
+
+static void
+on_enumerate_jit_method (MonoDomain *domain, MonoMethod *method, MonoJitInfo *jinfo, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumMethods++;
+
+	method_load (domain, method, jinfo, TRUE);
+}
+
+static void
+on_enumerate_domain (MonoDomain *domain, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumDomains++;
+
+	// Iterate through each assembly
+	mono_domain_assembly_foreach (domain, on_enumerate_assembly, enumerationData);
+
+	// Iterate through each JIT'ed method
+	mono_domain_jit_foreach (domain, on_enumerate_jit_method, enumerationData);
+}
+
+static void
+on_attach ()
+{
+	SUPERLUMINAL_LOG ("Enumerating JIT data...");
+
+	struct JITEnumerationData enumerationData;
+	memset (&enumerationData, 0, sizeof (enumerationData));
+
+	// Iterate through each domain
+	mono_domain_foreach (on_enumerate_domain, &enumerationData);
+
+	SUPERLUMINAL_LOG_ARGS ("Finished enumerating JIT data. Found %d domains, %d assemblies, %d methods", enumerationData.mNumDomains, enumerationData.mNumAssemblies, enumerationData.mNumMethods);
+}
+
+// This callback is called by the ETW system when tracing is started / stopped. We use it to enumerate & output information about JIT compilation that happened *before* tracing started.
+DECLSPEC_NOINLINE __inline VOID __stdcall Private_EventControlCallback (_In_ LPCGUID SourceId, _In_ ULONG ControlCode, _In_ UCHAR Level, _In_ ULONGLONG MatchAnyKeyword, _In_ ULONGLONG MatchAllKeyword, _In_opt_ PEVENT_FILTER_DESCRIPTOR FilterData, _Inout_opt_ PVOID CallbackContext)
+{
+	SUPERLUMINAL_LOG_ARGS ("EventControlCallback (%d)", ControlCode);
+
+	switch (ControlCode) {
+	case EVENT_CONTROL_CODE_ENABLE_PROVIDER: {
+		gboolean isRegular = (CallbackContext == &MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_Context);
+		gboolean isRundown = (CallbackContext == &MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_Context);
+		SUPERLUMINAL_LOG_ARGS("EventControlCallback -- IsInitialized: %s, IsRuntime: %s, IsRundown: %s", is_initialized ? "true" : "false", isRegular ? "true" : "false", isRundown ? "true" : "false");
+		if (is_initialized && CallbackContext == &MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_Context) {
+			on_attach ();
+		}
+	} break;
+	default:
+		break;
+	};
+}
+
+/* the entry point */
+MONO_API void
+mono_profiler_init_superluminal (const char *desc)
+{
+	SUPERLUMINAL_LOG ("Initializing Plugin");
+
+	EventRegisterMicrosoft_Windows_DotNETRuntime ();
+	EventRegisterMicrosoft_Windows_DotNETRuntimeRundown ();
+
+	// We currently need debug info to be able to read out the line number information, so force enable it here.
+	if (!mono_debug_enabled ())
+		mono_debug_init (MONO_DEBUG_FORMAT_MONO);
```

Comment:
What's the implication of this? Does this cause JIT to generate less than optimal code? Or is it just increased memory usage?

Reply:
AFAIK `mono_debug_init` just sets some flags and sets up some data structures to keep track of debugging information during JIT compilation and doesn't affect codegen in any way, so just increased memory indeed.

But this might be worth double-checking with somebody more at home in Mono as well.

Reply:
I can investigate it a bit. @joncham: Do you know off hand though?

Reply:
I'll have to check, but I think this just causes the loading of debug symbols when available and does store extra info to lookup file/line numbers, but should not result in JIT de-optimizations.